Client/game_sa/CDirectorySA.cpp

```diff
@@ -11,12 +11,73 @@
 #include "StdInc.h"
 #include "CDirectorySA.h"
 
+bool CDirectorySAInterface::AddEntry(DirectoryInfoSA& entry)
+{
+    if (m_numEntries >= m_capacity || GetModelEntry(entry.m_name))
+        return false;
+
+    entry.m_offset = 0;
+
+    if (m_numEntries > 0)
+    {
+        DirectoryInfoSA* lastEntry = m_entries + m_numEntries - 1;
+        entry.m_offset = lastEntry->m_offset + lastEntry->m_streamingSize;
+
+        if (entry.m_offset % 2048)
+            entry.m_offset += 2048 - (entry.m_offset % 2048);
+    }
+
+    m_entries[m_numEntries++] = entry;
+
+    return true;
+}
+
+bool CDirectorySAInterface::RemoveEntry(const char* fileName)
+{
+    if (m_numEntries <= 0)
+        return false;
+
+    DirectoryInfoSA* entry = GetModelEntry(fileName);
+
+    if (!entry)
+        return false;
+
+    std::ptrdiff_t index = entry - m_entries;
+
+    if (index < m_numEntries - 1)
+    {
+        DirectoryInfoSA* lastEntry = m_entries + m_numEntries - 1;
+        entry->m_offset = lastEntry->m_offset + lastEntry->m_sizeInArchive;
+    }
+
+    m_numEntries--;
+
+    if (index < m_numEntries)
+        std::memmove(entry, entry + 1, (m_numEntries - index) * sizeof(DirectoryInfoSA));
+
+    return true;
+}
+
+DirectoryInfoSA* CDirectorySAInterface::GetModelEntry(const char* fileName)
+{
+    if (m_numEntries <= 0)
```

Comment:
`m_numEntries` is unsigned. So maybe you meant `m_numEntries == 0`? The same for below.